aw_watcher_window/macos.swift

```diff
@@ -0,0 +1,285 @@
+
+@objc protocol ChromeProtocol {
+  @objc optional func windows() -> [ChromeWindow]
+}
+
+extension SBApplication: ChromeProtocol {}
+
+struct NetworkMessage: Codable {
+  var app: String
+  var title: String
+  var url: String?
+}
+
+struct Heartbeat: Codable {
+  var timestamp: Date
+  var data: NetworkMessage
+}
+
+struct Bucket: Codable {
+  var client: String
+  var type: String
+  var hostname: String
+}
+
+let clientHostname = ProcessInfo.processInfo.hostName
+let clientName = "aw-watcher-window2"
+let bucketName = "\(clientName)_\(clientHostname)"
+
+let main = MainThing()
+var oldHeartbeatData: NetworkMessage?
+
+let encoder = JSONEncoder()
+let formatter = ISO8601DateFormatter()
+formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
+
+encoder.dateEncodingStrategy = .custom({ date, encoder in
+  var container = encoder.singleValueContainer()
+  let dateString = formatter.string(from: date)
+  try container.encode(dateString)
+})
+
+start()
+RunLoop.main.run()
+
+func start() {
+  guard checkAccess() else {
+    DispatchQueue.main.asyncAfter(deadline: .now() + 10) {
+      start()
+    }
+    return
+  }
+
+  createBucket()
+
+  NSWorkspace.shared.notificationCenter.addObserver(
+    main, selector: #selector(main.focusedAppChanged),
+    name: NSWorkspace.didActivateApplicationNotification,
+    object: nil)
+  main.focusedAppChanged()
+  detectIdle()
+}
+
+func detectIdle() {
+  let seconds = 15.0 - SystemIdleTime()!
+  if seconds < 0.0 {
+    sendHeartbeat(data: NetworkMessage(app: "", title: ""))
+
+    var monitor: Any?
+    monitor = NSEvent.addGlobalMonitorForEvents(matching: [
+      .mouseMoved, .leftMouseDown, .rightMouseDown, .keyDown,
+    ]) { e in
+      NSEvent.removeMonitor(monitor!)
+      if let oldMenu = oldHeartbeatData { sendHeartbeat(data: oldMenu) }
+      detectIdle()
+    }
+
+    return
+  }
+
+  DispatchQueue.main.asyncAfter(deadline: .now() + seconds) {
+    detectIdle()
+  }
+}
+
+func createBucket() {
+  let payload = try! encoder.encode(
+    Bucket(client: clientName, type: "currentwindow", hostname: clientHostname))
+
+  let url = URL(string: "http://localhost:5600/api/0/buckets/\(bucketName)")!
+  Task {
+    var urlRequest = URLRequest(url: url)
+    urlRequest.httpMethod = "POST"
+    urlRequest.addValue("application/json", forHTTPHeaderField: "Content-Type")
+    let (_, response) = try await URLSession.shared.upload(for: urlRequest, from: payload)
+    guard (200...299).contains((response as! HTTPURLResponse).statusCode) else {
+      print("Failed to create bucket")
+      return
+    }
+  }
+}
+
+func sendHeartbeat(data: NetworkMessage) {
+  let payload0 = oldHeartbeatData.map {
+    try! encoder.encode(Heartbeat(timestamp: Date.now, data: $0))
+  }
+  let payload = try! encoder.encode(Heartbeat(timestamp: Date.now, data: data))
+
+  let url = URL(string: "http://localhost:5600/api/0/buckets/\(bucketName)/heartbeat?pulsetime=999999")!
```

Comment:
Don't set `pulsetime` to an arbitrarily high value, it will lead to incorrectly merging identical events in case they occur with a long empty interval between (for example, if the computer goes to sleep).

Set it to whatever the expected time is since the last known event (for polling, this is a value slightly higher than the poll interval). 

Since this doesn't use polling, you have to check that the last event was triggered somewhat recently, or combine it with fallback-polling if no event has been triggered recently (for example, in the last minute or so).

The other watchers work in a way such that ActivityWatch should receive an event even if the window title doesn't change (given enough time), adopting a "listen for changes, but poll if long since last check" approach would also help ensure this behavior stays consistent across platforms.

Reply:
Is there a websocket based api to use the automatic ping/pong to do keep alive?